aspnetcore/native-aot/index.md

```diff
@@ -0,0 +1,183 @@
+  Generating native code
+  MyFirstAotWebApi -> C:\Code\Demos\MyFirstAotWebApi\bin\Release\net8.0\win-x64\publish\
+```
+
+Note: The preceding output my differ from what you see depending on the version of  .NET 8 used.
+Review the contents of output directory:
+
+```
+$ dir bin\Release\net8.0\win-x64\publish
+
+    Directory: C:\Code\Demos\MyFirstAotWebApi\bin\Release\net8.0\win-x64\publish
+
+Mode                 LastWriteTime         Length Name
+----                 -------------         ------ ----
+-a---          30/03/2023  1:41 PM        9480704 MyFirstAotWebApi.exe
+-a---          30/03/2023  1:41 PM       43044864 MyFirstAotWebApi.pdb
+```
+
+The executable is self-contained and doesn't require a .NET runtime to run. When launched it should behave as expected:
+
+```
+$ .\bin\Release\net8.0\win-x64\publish\MyFirstAotWebApi.exe
+info: Microsoft.Hosting.Lifetime[14]
+      Now listening on: http://localhost:5000
+info: Microsoft.Hosting.Lifetime[0]
+      Application started. Press Ctrl+C to shut down.
+info: Microsoft.Hosting.Lifetime[0]
+      Hosting environment: Production
+info: Microsoft.Hosting.Lifetime[0]
+      Content root path: C:\Code\Demos\MyFirstAotWebApi
+```
+
+The `Program.cs` source file contains some changes for publishing to .NET native AOT. A significant difference is that `Microsoft.AspNetCore.Builder.WebApplication.CreateSlimBuilder` is used to create the web application builder.  The `CreateSlimBuilder` method initializes the <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder> with the minimal ASP.NET Core features necessary to run an application.
+<!-- Update the preceding with the following when the .NET 8 API is published :
+<xref:Microsoft.AspNetCore.Builder.WebApplication.CreateBuilderSlim%2A>
+-->
+
+```csharp
+var builder = WebApplication.CreateSlimBuilder(args);
+```
+
+Because this template uses JSON to serialize responses, we need to explicitly provide a `JsonSerializerContext` which specifies the custom types that we need to serialize so that the JSON source generator knows what to produce code for.
+
+```csharp
+builder.Services.ConfigureHttpJsonOptions(options =>
+{
+    options.SerializerOptions.AddContext<AppJsonSerializerContext>();
+});
+
+// Other code trimmed for brevity.
+
+[JsonSerializable(typeof(Todo[]))]
+internal partial class AppJsonSerializerContext : JsonSerializerContext
+{
+
+}
+```
+
+Because unused code is trimmed during publishing for Native AOT, the application cannot use unbounded reflection at runtime. Source generators are used to produce code to avoid the need for reflection. In some cases source generators produce code optimized for AOT even when a generator is not strictly required. To view source code that is generated based on the code in ```Program.cs``` modify the ```MyFirstAotWebApi.csproj``` to include the ```<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>``` property. Example:
```

Comment:
```suggestion
When using native AOT, unused code is trimmed during publishing to reduce the size of the final binary. This can break applications or libraries that rely on reflection at runtime. Source generators are used to produce code to avoid the need for reflection. In some cases source generators produce code optimized for AOT even when a generator is not strictly required. To view source code that is generated based on the code in ```Program.cs``` modify the ```MyFirstAotWebApi.csproj``` to include the ```<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>``` property. Example:
```

I would link to the runtime document that describes how to build trimmer friendly APIs and how to describe reflection.

Reply:
This was previously When ... but was changed for style guide reasons. Ack on link.

Reply:
I won't take this suggestion for now @davidfowl but I do think we need to some how point folks to docs about how to build trim-safe APIs in their own code. I'm not sure this is an ASP.NET specific problem though. I've got a link to the main native AOT landing page.

I think we we were going to have a discussion of trim safe APIs in our doc tree then we'd want to make it cover patterns that come up frequently in ASP.NET Core apps in particular. Maybe a sub-page in the TOC?